oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,212 @@
+
+Missing Translations
+====================
+
+Missing translations can be a disorienting experience for a user. We strongly suggest a careful evaluation of translation coverage, or a clear indicator of beta status, before making a new language available to the public.
+
+****************
+Server-Side i18n
+****************
+
+Have your server send symbolic values or error codes, rather than translated display text, to convey information, and then localize on the front end. This keeps all of your localization together, simplifies testing your server code, and makes your server response reusable by other front ends, which may wish to display the same information but with different text.
+
+#################
+Technology Choice
+#################
+
+**********
+react-intl
+**********
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/master/docs/decisions/0002-react-i18n.rst>`_. Our needs included the ability to add notes for translators and to get a translated message as a plain string (as opposed to only as a React component). react-intl is in line with current industry standards in i18n and meets edX needs. In order to add some browser shims, the Architecture team has published a thin wrapper around react-intl in the `frontend-i18n repo <https://github.com/edx/frontend-i18n>`_. For consistency, use this wrapper, instead of importing react-intl directly.
+
+Alternatives Considered
+=======================
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. It comes with plug-and-play functionality to determine a user's locale, babel scripts for various translation management schemes, and a lot of other supporting code. It also exposes a much cleaner interface for getting a translated message as a plain string. However, we would have had to write a custom solution for extracting translator comments, and it outputs a format that Transifex can't work with.
+
+************************
+Reference Implementation
+************************
+
+The reference implementation of this OEP is the Order History app in `frontend-app-ecommerce <https://github.com/edx/frontend-app-ecommerce>`_, which uses the `frontend-i18n <https://github.com/edx/frontend-i18n>`_ library.
+
+###########
+Future Work
+###########
+
+There is a special accessibility use case where a message id doesn't have a translation in the user's preferred locale, so we fall back to a default locale. In this case, the string would ideally be wrapped in a ``<span locale="....">`` tag to maintain the correct locale information for accessibility tools. In order to make sure that this string wrapping always happens, studio-frontend uses ``WrappedMessage``, a simple wrapper around the standard react-intl ``FormattedMessage`` component.
```

Comment:
Can you clarify?  Does this mean that `studio-frontend` handles this accessibility case, but `frontend-i18n` doesn't yet?